### ISSUE TITLE
MarkerClustering

### ISSUE THREAD

**Issue by AndrewAdams3** [googlemap-react/googlemap-react#34]
I think marker clustering would be a really nice addition to the library to have this kind of support. I am working on a project that needs to support clustering and has a constantly changing number of markers to render to the screen. I was able to accomplish it with a pretty simple solution:
```javascript
    useEffect(()=>{
      if(state.map){
        dispatch({type: "add_object", id: CLUSTER_ID, object: new MarkerClusterer(state.map, [], {
          maxZoom: 20,
          imagePath: 'https://developers.google.com/maps/documentation/javascript/examples/markerclusterer/m',
        })})
        setMapInit(true)
      }
    },[state.map])
    useEffect(()=>{
      if(state.map && mapInit && markers.length){
          const _ms = markers.map((marker)=>state.objects.has(marker.props.home._id) ? state.objects.get(marker.props.home._id) : null).filter((marker)=>!!marker)
          state.objects.get(CLUSTER_ID).clearMarkers()
          state.objects.get(CLUSTER_ID).addMarkers(_ms)
          state.objects.get(CLUSTER_ID).fitMapToMarkers()
      } else if(mapInit) {
        state.objects.get(CLUSTER_ID).clearMarkers()
      }
    },[markers.length, state.map, mapInit])
```
The only problem with this, however, is that each time the number of markers changes, the marker clusterer had to be completely reset and the new markers had to be added. Preferably, there would be a solution to only need to add or remove the markers that change.

My idea was to integrate a markerclusterer built into the mapbox and update the dispatch to handle adding / removing markers from the marker cluster so that the can be dealt with one at a time as they change. 

I have forked the repo and added the functionality I described, and everything seems to be working good. I added a `useClusterer` option and `clusterOpts` to mapbox, and I change the add_object and remove_object dispatch events to handle adding / removing markers from the clusterer.

The only issue I see is that I added the dependency on @google/markerclustererplus to give the base markerclusterer class. I assume implementing this on our own could be an option, but seeing as the dependency is small (only one file pretty much that defines the class) and is up to date (last edit was 6 days ago) I figured it wouldn't be much to worry about.

What do you think?

**Comment by lucifer1004** [comment]
I think maybe we can make a separate package to offer the capability, so that the core package will not have extra deps.




| |
吴自华
邮箱：qqbbnease1004@126.com
|

签名由 网易邮箱大师 定制

On 11/21/2019 02:29, Andrew Adams wrote:

I think marker clustering would be a really nice addition to the library to have this kind of support. I am working on a project that needs to support clustering and has a constantly changing number of markers to render to the screen. I was able to accomplish it with a pretty simple solution:

useEffect(()=>{
      if(state.map){
        dispatch({type:"add_object", id:CLUSTER_ID, object:newMarkerClusterer(state.map, [], {
          maxZoom:20,
          imagePath:'https://developers.google.com/maps/documentation/javascript/examples/markerclusterer/m',
        })})
        setMapInit(true)
      }
    },[state.map])
    useEffect(()=>{
      if(state.map&& mapInit &&markers.length){
          const_ms=markers.map((marker)=>state.objects.has(marker.props.home._id) ?state.objects.get(marker.props.home._id) :null).filter((marker)=>!!marker)
          state.objects.get(CLUSTER_ID).clearMarkers()
          state.objects.get(CLUSTER_ID).addMarkers(_ms)
          state.objects.get(CLUSTER_ID).fitMapToMarkers()
      } elseif(mapInit) {
        state.objects.get(CLUSTER_ID).clearMarkers()
      }
    },[markers.length, state.map, mapInit])

The only problem with this, however, is that each time the number of markers changes, the marker clusterer had to be completely reset and the new markers had to be added. Preferably, there would be a solution to only need to add or remove the markers that change.

My idea was to integrate a markerclusterer built into the mapbox and update the dispatch to handle adding / removing markers from the marker cluster so that the can be dealt with one at a time as they change.

I have forked the repo and added the functionality I described, and everything seems to be working good. I added a useClusterer option and clusterOpts to mapbox, and I change the add_object and remove_object dispatch events to handle adding / removing markers from the clusterer.

The only issue I see is that I added the dependency on @google/markerclustererplus to give the base markerclusterer class. I assume implementing this on our own could be an option, but seeing as the dependency is small (only one file pretty much that defines the class) and is up to date (last edit was 6 days ago) I figured it wouldn't be much to worry about.

What do you think?

**Comment by AndrewAdams3** [comment]
By extra package do you mean to write our own MarkerClusterer class instead of pulling from the @google/markerclustererplus package? If so, I could get that going. It would pretty much be a copy of that one but translated to typescript.

**Comment by lucifer1004** [comment]
No, I was thinking about put the wrapper for @google/markerclusterplus in a single package.
